p2p/discovery.py

```
@@ -337,33 +340,28 @@ def update_routing_table(self, node: NodeAPI) -> None:
-            neighbours.extend(response)
-            # This callback is expected to be called multiple times because nodes usually
-            # split the neighbours replies into multiple packets, so we only call event.set() once
-            # we've received enough neighbours.
-            if len(neighbours) >= constants.KADEMLIA_BUCKET_SIZE:
-                event.set()
-
-        with self.neighbours_callbacks.acquire(remote, process):
-            with trio.move_on_after(constants.KADEMLIA_REQUEST_TIMEOUT) as cancel_scope:
-                await event.wait()
-                self.logger.debug2('got expected neighbours response from %s', remote)
-            if cancel_scope.cancelled_caught:
-                self.logger.debug2(
-                    'timed out waiting for %d neighbours from %s',
-                    constants.KADEMLIA_BUCKET_SIZE,
-                    remote,
-                )
-
+        send_chan, recv_chan = trio.open_memory_channel[List[NodeAPI]](1)
+        try:
+            with trio.fail_after(constants.KADEMLIA_REQUEST_TIMEOUT):
+                # Responses to a FIND_NODE request are usually split between multiple
+                # NEIGHBOURS packets, so we may have read from the channel multiple times.
+                gen = self.neighbours_channels.receive(remote, send_chan, recv_chan)
+                # mypy thinks wrapping our generator turns it into something else, so ignore.
+                async with aclosing(gen):  # type: ignore
+                    async for batch in gen:
+                        self.logger.debug2(
+                            f'got expected neighbours response from {remote}: {batch}')
+                        neighbours.extend(batch)
+                        if len(neighbours) >= constants.KADEMLIA_BUCKET_SIZE:
+                            break
+                self.logger.debug2(f'got expected neighbours response from {remote}')
+        except trio.TooSlowError:
```

I think the following format might be more readable.

```python
with trio.move_on_after(...) as cancel_scope:
    ...
if cancel_scope.cancelled_caught:
    self.logger.debug(...)
```

Just a slight reduction in levels of indention.  It does have the slight downside of allowing for the silent failure if the `if cancel_scope.cancelled_caught` block is accidentally removed, but I think I prefer this over the large `try` block.

---

Yeah, I think I also prefer this over the large try and extra level of indentation